src/JustSaying/Fluent/PublishConfig/DynamicMessagePublisher.cs

```diff
@@ -0,0 +1,74 @@
+using System.Collections.Concurrent;
+using JustSaying.Messaging;
+using JustSaying.Messaging.Interrogation;
+using JustSaying.Models;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Fluent;
+
+internal class DynamicMessagePublisher : IMessagePublisher
+{
+    private readonly ConcurrentDictionary<string, IMessagePublisher> _publisherCache = new();
+    private readonly Func<Message, string> _topicNameCustomizer;
+    private readonly Func<string, StaticPublicationConfiguration> _staticConfigBuilder;
+
+    private readonly ConcurrentDictionary<string, SemaphoreSlim> _topicCreationLocks = new();
+    private readonly ILogger<DynamicMessagePublisher> _logger;
+
+    public DynamicMessagePublisher(Func<Message, string> topicNameCustomizer, Func<string, StaticPublicationConfiguration> staticConfigBuilder, ILoggerFactory loggerFactory)
+    {
+        _topicNameCustomizer = topicNameCustomizer;
+        _staticConfigBuilder = staticConfigBuilder;
+        _logger = loggerFactory.CreateLogger<DynamicMessagePublisher>();
+    }
+
+    public InterrogationResult Interrogate()
+    {
+        var pairs = _publisherCache.Keys.OrderBy(x => x)
+            .ToDictionary(x => x, x => _publisherCache[x].Interrogate());
```

Comment:
Would this operation be thread-safe between accessing the keys and getting the value from the cache?

Reply:
No it wouldn't be, although worst case there might be a new publisher that isn't captured in the `Keys` so it wouldn't be emitted. We could add a lock mechanism around it which would block publishers during interrogation, but given it's just interrogation I'm not sure it's worth it? The member accesses themselves are all thread safe so it's just a behaviour thing.

What do you think?